src/lib.rs

```diff
@@ -290,49 +288,28 @@ impl Iterator for Masks {
 
 
 pub struct Ones<'a> {
-    current_bit_idx: usize,
-    current_block_idx: usize,
+    bitset: Block,
+    block_idx: usize,
     remaining_blocks: &'a [Block],
-    current_block: Block
 }
 
 impl<'a> Iterator for Ones<'a> {
     type Item = usize; // the bit position of the '1'
 
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        let mut block = self.current_block;
-        let mut idx = self.current_bit_idx;
-
-        loop {
-            loop {
-                if (block & 1) == 1 {
-                    self.current_block = block >> 1;
-                    self.current_bit_idx = idx + 1;
-                    return Some(idx);
-                }
-                // reordering the two lines below makes a huge (2x) difference in performance!
-                block = block >> 1;
-                idx += 1;
-                if block == 0 {
-                    break;
-                }
-            }
-
-            // go to next block
-            match self.remaining_blocks.split_first() {
-                Some((&next_block, rest)) => {
-                    self.remaining_blocks = rest;
-                    self.current_block_idx += 1;
-                    idx = self.current_block_idx * BITS;
-                    block = next_block;
-                }
-                None => {
-                    // last block => done
-                    return None;
-                }
+        while self.bitset == 0 {
+            if self.remaining_blocks.is_empty() {
+                return None;
             }
+            self.bitset = self.remaining_blocks[0];
```

Comment:
Instead of setting bitset to remaing_blocks[0] and updating remaining blocks, why dont you search for the first none zero block and only then update remaining_blocks?

Reply:
(the PR has been opened for 2 years, so I guess nothing needs be done until the maintainer shows interest, but I figured I'd add the comment in case something comes of it)

Reply:
I'm not sure that this is faster, because it might mess up vectorization. Could be interesting to benchmark.